CenterForOpenScience/osf.io
# [Feature][Ready for Review] Personal Access Tokens

Purpose
======
Add view for a (depth / API) user to generate personal access tokens. (see [github's implementation](https://github.com/settings/tokens) for further explanation)

Changes
=======
* Model added
* View added

Side Effects
=========
None

Note
-------
* CAS changes went in already (h/t @icereval )

Testing
----------
1) Create a private project (and make note of the `pid`)
2) Create (and make note of the `token_id` for) a personal access token
3) With cas running:
 * `curl -v -H "Authorization: Bearer [token_id here]" -X GET -k https://127.0.0.1:8443/oauth2/profile`
 * `curl -v -H "Authorization: Bearer [token_id here]" -X GET http://127.0.0.1:8000/v2/nodes/[pid here]/`

3.1) Sample expected responses:
```bash
~/D/osf.io ❯ curl -v -H "Authorization: Bearer d98974df78c649ae85d520c07c1379fa" -X GET -k https://127.0.0.1:8443/oauth2/profile
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
* Server certificate: 127.0.0.1
> GET /oauth2/profile HTTP/1.1
> Host: 127.0.0.1:8443
> User-Agent: curl/7.43.0
> Accept: */*
> Authorization: Bearer d98974df78c649ae85d520c07c1379fa
>
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=UTF-8
< Content-Length: 57
<
* Connection #0 to host 127.0.0.1 left intact
{"scope":"osf.full_write osf.full_read","id":"hpn4b"}%

~/D/osf.io ❯ curl -v -H "Authorization: Bearer d98974df78c649ae85d520c07c1379fa" -X GET  http://127.0.0.1:8000/v2/nodes/f4h36/
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> GET /v2/nodes/f4h36/ HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/7.43.0
> Accept: */*
> Authorization: Bearer d98974df78c649ae85d520c07c1379fa
>
< HTTP/1.1 200 OK
< Vary: Accept
< X-Frame-Options: SAMEORIGIN
< Content-Type: application/vnd.api+json
< Allow: GET, PUT, PATCH, DELETE, HEAD, OPTIONS
< Content-Length: 1044
< Server: livereload-tornado
<
* Connection #0 to host 127.0.0.1 left intact
{"data":{"id":"f4h36","type":"nodes","attributes":{"title":"testproj","description":"","category":"project","date_created":"2015-10-20T15:26:12.149000","date_modified":"2015-10-20T15:26:12.149000","registration":false,"fork":false,"collection":false,"dashboard":false,"tags":[],"public":false},"relationships":{"children":{"links":{"related":{"href":"http://127.0.0.1:8000/v2/nodes/f4h36/children/","meta":{}}}},"contributors":{"links":{"related":{"href":"http://127.0.0.1:8000/v2/nodes/f4h36/contributors/","meta":{}}}},"files":{"links":{"related":{"href":"http://127.0.0.1:8000/v2/nodes/f4h36/files/","meta":{}}}},"node_links":{"links":{"related":{"href":"http://127.0.0.1:8000/v2/nodes/f4h36/node_links/","meta":{}}}},"parent":{"links":{"related":{"href":null,"meta":{}}}},"registrations":{"links":{"related":{"href":"http://127.0.0.1:8000/v2/nodes/f4h36/registrations/","meta":{}}}},"forked_from":{"links":{"related":{"href":null,"meta":{}}}}},"links":{"self":"http://localhost:8000/v2/nodes/f4h36/","html":"http://localhost:5000/f4h36/"}}}%
```

4) Deauthorize token through UI
5) Try to hit either endpoint above again

5.1) Expected response:
```bash
~/D/osf.io ❯ curl -v -H "Authorization: Bearer d98974df78c649ae85d520c07c1379fa" -X GET -k https://127.0.0.1:8443/oauth2/profile
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
* Server certificate: 127.0.0.1
> GET /oauth2/profile HTTP/1.1
> Host: 127.0.0.1:8443
> User-Agent: curl/7.43.0
> Accept: */*
> Authorization: Bearer d98974df78c649ae85d520c07c1379fa
>
< HTTP/1.1 401 Unauthorized
< Content-Type: application/json; charset=UTF-8
< Content-Length: 67
<
* Connection #0 to host 127.0.0.1 left intact
{"error_description":"Invalid Access Token","error":"unauthorized"}%
```

6) Try to hit an endpoint that a token doesn't have access to
6.1) In this case, a token with `osf.users.all_read` doesn't have access to view a node:
```bash
~/D/osf.io ❯ curl -v -H "Authorization: Bearer f04afa93c7b2473780a46624d299ecf0" -X GET http://127.0.0.1:8000/v2/nodes/f4h36/
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> GET /v2/nodes/f4h36/ HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/7.43.0
> Accept: */*
> Authorization: Bearer f04afa93c7b2473780a46624d299ecf0
>
< HTTP/1.1 403 FORBIDDEN
< Vary: Accept
< X-Frame-Options: SAMEORIGIN
< Content-Type: application/vnd.api+json
< Allow: GET, PUT, PATCH, DELETE, HEAD, OPTIONS
< Content-Length: 76
< Server: livereload-tornado
<
* Connection #0 to host 127.0.0.1 left intact
{"errors":[{"detail":"You do not have permission to perform this action."}]}%
```

Associated UI
-------------------

List view:
<img width="986" alt="screen shot 2015-10-23 at 12 14 23" src="https://cloud.githubusercontent.com/assets/5659262/10698628/011852ea-7980-11e5-9b65-d54e0832d30a.png">

Create:
<img width="575" alt="screen shot 2015-10-23 at 12 14 38" src="https://cloud.githubusercontent.com/assets/5659262/10698631/011b9e5a-7980-11e5-86f8-c76415730cc6.png">

Just Created:
<img width="800" alt="screen shot 2015-10-23 at 12 15 32" src="https://cloud.githubusercontent.com/assets/5659262/10698629/011976d4-7980-11e5-95c2-d8e66de6c4c0.png">

Edit existing (note lack of `token_id`):
<img width="698" alt="screen shot 2015-10-23 at 12 15 56" src="https://cloud.githubusercontent.com/assets/5659262/10698630/011aa68a-7980-11e5-9c6b-f94060e29dd5.png">









[#OSF-3138]
